### PR TITLE
reducedim special handling of Complex element types. fix #28320

### LINKED ISSUE #28320: mapreducedim issue with abstract complex type
When calling `mapreduce(abs2, +, Complex[1.0 x], dims=2)` we obtain unexpected behavior.
- if `x` has an integer value, the result type is `Array{Int}` it should be `Array{Float64}`.
- if `x` cannot be converted to an `Int` an exception is thrown

```
julia> A = Complex[1.0 2.0]
1×2 Array{Complex,2}:
 1.0+0.0im  2.0+0.0im

julia> mapreduce(abs2, +, A, dims=2)
1×1 Array{Int64,2}:
 5

julia> A = Complex[1.0 2.5]
1×2 Array{Complex,2}:
 1.0+0.0im  2.5+0.0im

julia> mapreduce(abs2, +, A, dims=2)
ERROR: InexactError: Int64(Int64, 7.25)
Stacktrace:
 [1] Type at ./float.jl:700 [inlined]
 [2] convert(::Type{Int64}, ::Float64) at ./number.jl:7
 [3] setindex! at ./array.jl:771 [inlined]
 [4] setindex! at ./multidimensional.jl:414 [inlined]
 [5] macro expansion at ./simdloop.jl:65 [inlined]
 [6] _mapreducedim!(::typeof(abs2), ::typeof(+), ::Array{Int64,2}, ::Array{Complex,2}) at ./reducedim.jl:249
 [7] mapreducedim!(::Function, ::Function, ::Array{Int64,2}, ::Array{Complex,2}) at ./reducedim.jl:265
 [8] _mapreduce_dim(::Function, ::Function, ::NamedTuple{(),Tuple{}}, ::Array{Complex,2}, ::Int64) at ./reducedim.jl:304
 [9] #mapreduce#542 at ./reducedim.jl:295 [inlined]
 [10] (::getfield(Base, Symbol("#kw##mapreduce")))(::NamedTuple{(:dims,),Tuple{Int64}}, ::typeof(mapreduce), ::Function, ::Function, ::Array{Complex,2}) at ./none:0
 [11] top-level scope at none:0

julia> versioninfo()
Julia Version 0.7.0-beta2.143
Commit cb6f5e225d (2018-07-28 00:48 UTC)
```
Notes:
- The same happens for `sum, prod, maximum, ...`.
- The same happens for sparse arrays
- No problem if I replace `abs2` with `abs`

I try to come up with a fixing PR soon.

### PR DESCRIPTION
**Pull request by KlausC**
Fix https://github.com/JuliaLang/julia/issues/28320 .

### CODE REVIEW

File: base/reducedim.jl [hunk 1]
```diff
@@ -99,14 +99,20 @@ reducedim_initarray(A::AbstractArray, region, init::T) where {T} = reducedim_ini
...
 function reducedim_init(f, op::Union{typeof(+),typeof(add_sum)}, A::AbstractArray, region)
     _reducedim_init(f, op, zero, sum, A, region)
 end
 function reducedim_init(f, op::Union{typeof(*),typeof(mul_prod)}, A::AbstractArray, region)
     _reducedim_init(f, op, one, prod, A, region)
 end
 function _reducedim_init(f, op, fv, fop, A, region)
-    T = promote_union(eltype(A))
+    T = _realtype(f, promote_union(eltype(A)))
     if T !== Any && applicable(zero, T)
```
Comment (Keno) [inline base/reducedim.jl]: how about instead a condition like `&& typeof(zero(T)) == T` here?

**Comment by Keno** [comment]
Also #10533 seems related.

**Comment by KlausC** [comment]
This PR tries to fix another corner case of a more general problem with minimal changes. It is closely related to https://github.com/JuliaLang/julia/pull/28260, which solved the same issue for arrays of abstract type `Real` which has been merged into mater by @JeffBezanson recently.
Issue  https://github.com/JuliaLang/julia/issues/10533, which has no solution yet, IMHO targets more generally the heuristics used in `reduce` (and `reducedim`) used to find
1. the element type of the container used to accumulate along a region (type of image of `f`)
2. a neutral element of the reducing function `fop` as initial value for the container elements
This problem seems not easy to solve, as the author of the original code claims
```
# TODO: better way to handle reducedim initialization
# The current scheme is basically following Steven G. Johnson's original implementation
```

**Comment by andreasnoack** [comment]
I'll merge this since it improves the current situation but we should really try to come up with a better approach here. Right now, the initialization mechanism is really ad hoc and sometimes extremely inefficient as in https://github.com/JuliaLang/julia/blob/0601b38267b0f467df681794c682cd3f708f2bad/base/reducedim.jl#L115 which computes the full reduction of the array before computing the reduction along dimensions.

**Pull request merged**